tests/src/test/scala/org/apache/openwhisk/core/scheduler/container/test/ContainerManagerTests.scala

```diff
@@ -1155,6 +1171,285 @@ class ContainerManagerTests
+        FullyQualifiedEntityName(EntityPath("ns1"), EntityName(testAction)),
+        testRevision,
+        actionMetadata,
+        testsid,
+        schedulerHost,
+        rpcPort)
+    val msg2 =
+      ContainerCreationMessage(
+        TransactionId.testing,
+        testInvocationNamespace,
+        FullyQualifiedEntityName(EntityPath("ns3"), EntityName(testAction)),
+        testRevision,
+        actionMetadata,
+        testsid,
+        schedulerHost,
+        rpcPort)
+    val msg3 =
+      ContainerCreationMessage(
+        TransactionId.testing,
+        testInvocationNamespace,
+        FullyQualifiedEntityName(EntityPath("ns3"), EntityName(testAction)),
+        testRevision,
+        actionMetadata,
+        testsid,
+        schedulerHost,
+        rpcPort)
+
+    val msgs = List(msg1, msg2, msg3)
+
+    val (coldCreations, warmedCreations) =
+      ContainerManager.filterWarmedCreations(warmedContainers, inProgressWarmedContainers, invokers, msgs)
+
+    warmedCreations.isEmpty shouldBe true
+    coldCreations.size shouldBe 3
+  }
+
+  it should "filter warmed creations when there are warmed containers" in {
+    val warmedContainers = Set(
+      ContainerKeys.warmedContainers(
+        testInvocationNamespace,
+        testfqn,
+        testRevision,
+        InvokerInstanceId(0, userMemory = 0.bytes),
+        ContainerId("fake")),
+      ContainerKeys.warmedContainers(
+        testInvocationNamespace,
+        testfqn.copy(name = EntityName("test-action-2")),
+        testRevision,
+        InvokerInstanceId(1, userMemory = 0.bytes),
+        ContainerId("fake")))
+    val inProgressWarmedContainers = TrieMap.empty[String, String]
+
+    val msg1 =
+      ContainerCreationMessage(
+        TransactionId.testing,
+        testInvocationNamespace,
+        testfqn,
+        testRevision,
+        actionMetadata,
+        testsid,
+        schedulerHost,
+        rpcPort)
+    val msg2 =
+      ContainerCreationMessage(
+        TransactionId.testing,
+        testInvocationNamespace,
+        testfqn.copy(name = EntityName("test-action-2")),
+        testRevision,
+        actionMetadata,
+        testsid,
+        schedulerHost,
+        rpcPort)
+    val msg3 =
+      ContainerCreationMessage(
+        TransactionId.testing,
+        testInvocationNamespace,
+        testfqn,
+        testRevision,
+        actionMetadata,
+        testsid,
+        schedulerHost,
+        rpcPort)
+
+    val msgs = List(msg1, msg2, msg3)
+
+    val (coldCreations, warmedCreations) =
+      ContainerManager.filterWarmedCreations(warmedContainers, inProgressWarmedContainers, invokers, msgs)
+
+    warmedCreations.size shouldBe 2
+    coldCreations.size shouldBe 1
+
+    warmedCreations.map(_._1).contains(msg1) shouldBe true
+    warmedCreations.map(_._1).contains(msg2) shouldBe true
+    coldCreations.map(_._1).contains(msg3) shouldBe true
+  }
+
+  it should "choose cold creation when warmed containers are in disabled invokers" in {
```

Comment:
This test will guarantee no disabled invoker is used because of warmed containers.